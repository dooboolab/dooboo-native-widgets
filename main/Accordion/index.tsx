import React, { FC } from 'react';
import AccrordionItem from './AccordionItem';
import styled from 'styled-components/native';

const Container = styled.ScrollView`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

type ItemType = {
  itemTitle: string;
  itemBodies: Array<string>;
}

interface Props {
  data: Array<ItemType>;
  isAnimated?: boolean;
  collapsedWhenRedered: boolean;
  animDuration?: number;
  activeOpacity?: number;
}

const Accordion: FC<Props> = (props) => {
  const {
    data,
    isAnimated,
    collapsedWhenRedered,
    animDuration,
    activeOpacity,
  } = props;

  return (
    <Container>
      {
        data.map((itemData, idx) => {
          return (
            <AccrordionItem
              key={idx}
              itemData={itemData}
              isAnimated={isAnimated}
              collapsedWhenRedered={collapsedWhenRedered}
              animDuration={animDuration}
              activeOpacity={activeOpacity}
            />
          );
        })
      }
    </Container>
  );
};

export default Accordion;
