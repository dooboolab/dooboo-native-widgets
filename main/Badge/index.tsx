import React, { FC } from 'react';

import { StyleSheet } from 'react-native';
import styled from 'styled-components/native';

interface StyleProps {
  color?: string;
  opacity?: number;
  badgePlacement?: 'left' | 'top' | 'bottom' | 'right';
  border?: string;
  textColor?: string;
}

export interface BadgeProps extends StyleProps{
  count?: number;
  maximumCount?: number;
  showZero?: boolean;
  opacityVisible?: boolean;
  variant?: string;
}

const StyledView = styled.View<StyleProps>`
  position: absolute;
  top: -15px;
  ${(props) => props.badgePlacement}: -10px;
  width: auto;
  min-width: 45px;
  height: 45px;
  border-color: ${(props) =>
    props.border ? props.border : '#00ff0000'};
  border-width: 3px;
  background-color: ${(props) => props.color};
  border-radius: 50px;
  justify-content: center;
  align-items: center;
  opacity: ${(props) => props.opacity};
`;

const StyledText = styled.Text<StyleProps>`
  color: ${(props) => props.textColor};
  text-align: center;
  padding: 5px;
  margin-left: 3px;
  margin-right: 3px;
`;

const StyledDotView = styled.View<StyleProps>`
  position: absolute;
  top: -5px;
  ${(props) => props.badgePlacement}: -5px;
  width: 20px;
  height: 20px;
  background-color: ${(props) => props.color};
  border-radius: 50px;
  justify-content: center;
  align-items: center;
`;

const styles = StyleSheet.create({
  fontDesign: {
    fontWeight: '500',
    fontSize: 20,
  },
});

const Badge: FC<BadgeProps> = (props) => {
  const {
    count = 1,
    color = '#34AFF9',
    maximumCount = 99,
    showZero,
    opacityVisible = true,
    variant = 'standard',
    badgePlacement: position = 'right',
    border,
    textColor = '#FFFFFF',
  } = props;

  if (!showZero) {
    if (count === 0) return null;
  }

  if (variant === 'dot') {
    return <StyledDotView color={color} badgePlacement={position} />;
  } else {
    const opacity = (maximumCount < count && opacityVisible) ? 0.6 : 1;
    const countText = (maximumCount < count) ? count + '+' : count;

    return (
      <StyledView
        color={color}
        opacity={opacity}
        badgePlacement={position}
        border={border}>
        <StyledText textColor={textColor} style={styles.fontDesign}>
          {countText}
        </StyledText>
      </StyledView>
    );
  }
};

export { Badge };
