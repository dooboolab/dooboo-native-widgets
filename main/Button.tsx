import {
  ActivityIndicator,
  LayoutRectangle,
  StyleProp,
  StyleSheet,
  Text,
  TextProps,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle,
} from 'react-native';
import React, {useRef, useState} from 'react';

import {useHover} from 'react-native-web-hooks';

const defaultStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    paddingHorizontal: 12,
    paddingVertical: 8,
    borderRadius: 6,

    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    fontSize: 14,
    color: '#069ccd',
  },
  disabledButton: {
    backgroundColor: '#cccccc',
    borderColor: 'rgb(200, 200, 200)',
  },
  disabledText: {
    color: '#969696',
  },
  hovered: {
    shadowColor: 'black',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.24,
    shadowRadius: 16,
    elevation: 10,
  },
});

type StylesType = Partial<StyleSheet.NamedStyles<typeof defaultStyles>>;

interface Props {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: StylesType;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
}

function Button({
  testID,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = '#ffffff',
  leftElement,
  rightElement,
  activeOpacity = 0.7,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
}: Props): React.ReactElement {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();

  return (
    <TouchableOpacity
      testID={testID}
      ref={ref}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <View
          style={[
            defaultStyles.container,
            styles?.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && [defaultStyles.hovered, styles?.hovered],
            disabled && [defaultStyles.disabledButton, styles?.disabledButton],
          ]}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </View>
      ) : (
        <View
          style={[
            defaultStyles.container,
            styles?.container,
            hovered && !disabled && [defaultStyles.hovered, styles?.hovered],
            disabled && [defaultStyles.disabledButton, styles?.disabledButton],
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}>
          {leftElement}
          <Text
            style={[
              defaultStyles.text,
              styles?.text,
              disabled && [defaultStyles.disabledText, styles?.disabledText],
            ]}
            {...textProps}>
            {text}
          </Text>
          {rightElement}
        </View>
      )}
    </TouchableOpacity>
  );
}

export {Button};
