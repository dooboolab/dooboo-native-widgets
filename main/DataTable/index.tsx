import * as React from 'react';

import {
  ScrollView,
  StyleProp,
  StyleSheet,
  Text,
  View,
  ViewStyle,
} from 'react-native';

import CheckBox from './CheckBox';
import DataTableCell from './DataTableCell';
import DataTableHeader from './DataTableHeader';
import DataTableRow from './DataTableRow';
import DataTableTitle from './DataTableTitle';
import { isHasValue } from './utils';
import styled from 'styled-components/native';

export interface Props {
  style?: StyleProp<ViewStyle>;
  data: Array<Record<string, any>>;
  isCheckAble?: boolean;
  customGroup?: Array<string>;
}

type DataTableNamespace = {
  Title: typeof DataTableTitle;
  Header: typeof DataTableHeader;
  Row: typeof DataTableRow;
  Cell: typeof DataTableCell;
};

const Container = styled.View`
  flex: 1;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
`;

const DataTable: React.FC<Props> & DataTableNamespace = ({
  data,
  isCheckAble,
  customGroup,
  style,
  ...rest
}) => {
  /** default Data key */
  const group = data.reduce((acc, current) => {
    Array.from(Object.assign(acc, Object.keys(current)));
    return acc;
  }, []);

  /** checking interaction */
  const [selected, setSelected] = React.useState<string[]>([]);
  const isSelected = (name: string): boolean => selected.indexOf(name) !== -1;
  const handleClick = (name: string): void => {
    const selectedIndex = selected.indexOf(name);
    let newSelected: string[] = [];
    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, name);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1),
      );
    }
    setSelected(newSelected);
  };
  /** Render */
  return (
    <ScrollView
      style={{
        width: '100%',
        height: '100%',
      }}>
      <ScrollView horizontal>
        <Container {...rest} style={[style]}>
          <View>
            <Text>
              {selected.length > 0 ? `${selected.length} item selected` : null}{' '}
            </Text>
          </View>
          <DataTable.Header style={{ backgroundColor: '#f9f9f9' }}>
            <DataTable.Title style={[!isCheckAble && styles.short]} />

            {/** have a customGroup or undefined  */}
            {(customGroup || group)?.map((field: any, index: number) => {
              return (
                <DataTable.Title
                  numberOfLines={1}
                  key={`${field}-${index}`}
                  style={[styles.center, index === 0 ? styles.short : null]}>
                  {field}
                </DataTable.Title>
              );
            })}
          </DataTable.Header>
          {data.map((item, i) => {
            const isItemSelected = isSelected(item[group[0]]);
            return (
              <DataTable.Row
                key={`row-${item}-${i}`}
                isChecked={!!isItemSelected}
                style={{ backgroundColor: 'white' }}>
                {/* If CheckAble is true */}
                {isCheckAble ? (
                  <DataTable.Cell style={[{ justifyContent: 'center' }]}>
                    <CheckBox
                      onClick={(): void => handleClick(item[group[0]])}
                      checked={!!isItemSelected}
                    />
                  </DataTable.Cell>
                ) : (
                  <DataTable.Cell style={[!isCheckAble && styles.short]} />
                )}
                {/** Body */}
                {group?.map((param: any, index: number) => {
                  return (
                    <DataTable.Cell
                      key={`cell-${param}-${index}`}
                      style={[
                        styles.center,
                        index === 0 ? styles.short : null,
                      ]}>
                      {isHasValue(item[param])}
                    </DataTable.Cell>
                  );
                })}
              </DataTable.Row>
            );
          })}
        </Container>
      </ScrollView>
    </ScrollView>
  );
};

DataTable.Title = DataTableTitle;
DataTable.Header = DataTableHeader;
DataTable.Row = DataTableRow;
DataTable.Cell = DataTableCell;

const styles = StyleSheet.create({
  short: {
    width: 50,
  },
  center: {
    justifyContent: 'center',
  },
});

export default DataTable;
