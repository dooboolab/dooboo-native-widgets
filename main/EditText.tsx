import {Platform, StyleSheet, Text, TextInput, View} from 'react-native';
import React, {FC, useState} from 'react';
import type {StyleProp, TextInputProps, ViewStyle} from 'react-native';

const defaultStyle: ViewStyle = {
  alignSelf: 'stretch',
  flex: 1,

  flexDirection: 'column',
};

const defaultRowStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    borderBottomWidth: 0.3,
    borderBottomColor: '#666666',
  },
  labelText: {
    fontSize: 14,
    color: '#b9b9c4',
    width: 100,
  },
  input: {
    paddingVertical: 12,
    paddingHorizontal: 8,
    fontSize: 14,
    fontWeight: 'bold',
    flex: 1,
    color: '#2c374e',
  },
  errorText: {
    marginTop: 4,
    fontSize: 12,
    color: '#991111',
  },
});

const defaultColumnStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    justifyContent: 'space-between',
    borderBottomWidth: 0.3,
    borderBottomColor: '#666666',

    flexDirection: 'column',
  },
  labelText: {
    fontSize: 14,
    color: '#b9b9c4',
  },
  input: {
    paddingVertical: 12,
    fontSize: 14,
    fontWeight: 'bold',
    flex: 1,
    color: '#2c374e',
  },
  errorText: {
    marginTop: 8,
    fontSize: 12,
    color: '#991111',
  },
});

type StylesType = Partial<StyleSheet.NamedStyles<typeof defaultRowStyles>>;

type Props = {
  testID?: TextInputProps['testID'];
  textInputProps?: TextInputProps;
  style?: StyleProp<ViewStyle>;
  styles?: StylesType;
  labelText?: string;
  errorText?: string;
  value?: TextInputProps['value'];
  onChange?: TextInputProps['onChange'];
  onChangeText?: TextInputProps['onChangeText'];
  placeholder?: TextInputProps['placeholder'];
  placeholderTextColor?: TextInputProps['placeholderTextColor'];
  onFocus?: TextInputProps['onFocus'];
  onBlur?: TextInputProps['onBlur'];
  editable?: TextInputProps['editable'];
  autoCapitalize?: TextInputProps['autoCapitalize'];
  secureTextEntry?: TextInputProps['secureTextEntry'];
  onSubmitEditing?: TextInputProps['onSubmitEditing'];
  focusColor?: string;
  errorColor?: string;
  disableColor?: string;
  type?: 'row' | 'column';
};

const EditText: FC<Props> = ({
  testID,
  textInputProps,
  style,
  styles,
  labelText = '',
  errorText = '',
  value = '',
  placeholder,
  placeholderTextColor,
  onChange,
  onChangeText,
  onFocus,
  onBlur,
  autoCapitalize = 'none',
  secureTextEntry = false,
  editable = true,
  focusColor = '#9A77FF',
  errorColor = '#ff7676',
  disableColor = '#999999',
  type = 'column',
}) => {
  const [focused, setFocused] = useState(false);

  const defaultStyles = type === 'row' ? defaultRowStyles : defaultColumnStyles;

  return (
    <View style={[defaultStyle, style]}>
      <View
        style={[
          defaultStyles.container,
          {
            borderColor: errorText
              ? errorColor
              : focused
              ? focusColor
              : disableColor,
            borderBottomColor: errorText
              ? errorColor
              : focused
              ? focusColor
              : disableColor,
          },
        ]}>
        {/* {labelText ? ( */}
        <Text
          style={[
            defaultStyles.labelText,
            styles?.labelText,
            {
              color: errorText
                ? errorColor
                : focused
                ? focusColor
                : disableColor,
            },
          ]}>
          {labelText}
        </Text>
        {/* ) : null} */}
        <TextInput
          {...textInputProps}
          testID={testID}
          autoCapitalize={autoCapitalize}
          secureTextEntry={secureTextEntry}
          style={[
            defaultStyles.input,
            styles?.input,
            // @ts-ignore
            Platform.OS === 'web' && {outlineWidth: 0},
          ]}
          editable={editable}
          onFocus={(e) => {
            setFocused(true);
            onFocus?.(e);
          }}
          onBlur={(e) => {
            setFocused(false);
            onBlur?.(e);
          }}
          value={value}
          placeholder={placeholder}
          placeholderTextColor={placeholderTextColor}
          onChange={onChange}
          onChangeText={onChangeText}
        />
      </View>
      {errorText ? (
        <Text
          style={[
            defaultStyles.errorText,
            styles?.errorText,
            {color: errorColor},
          ]}>
          {errorText}
        </Text>
      ) : null}
    </View>
  );
};

export {EditText};
