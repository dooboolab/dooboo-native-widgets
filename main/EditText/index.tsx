import {
  KeyboardTypeOptions,
  NativeSyntheticEvent,
  Platform,
  TextInputProps,
  TextInputSubmitEditingEventData,
  TextStyle,
  ViewStyle,
} from 'react-native';
import React, { FC, ReactElement, useState } from 'react';

import styled from 'styled-components/native';

const StyledRowContent = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

const StyledRowLabel = styled.Text`
  font-size: 14px;
  font-weight: bold;
  color: #b9b9c4;
`;

const StyledRowInput = styled.TextInput`
  padding-top: 16px;
  padding-bottom: 16px;
  font-size: 14px;
  font-weight: bold;
  flex: 1;
  height: 100%;
  color: #2c374e;
  ${Platform.OS === 'web' && { 'outline-style': 'none' }}
`;

const Container = styled.View`
  flex-direction: column;
  align-self: stretch;
`;

const StyledLine = styled.View`
  flex-direction: row;
  align-items: center;
  width: 100%;
`;

const StyledLabel = styled.Text`
  margin-bottom: 5px;
  font-size: 14px;
  font-weight: 500;
  color: #b9b9c4;
`;

const StyledIcon = styled.View`
  align-items: center;
  justify-content: center;
`;

const StyledTextInput = styled.TextInput`
  padding-top: 15px;
  padding-bottom: 15px;
  font-size: 15px;
  font-weight: 500;
  flex: 1;
  ${Platform.OS === 'web' && { 'outline-style': 'none' }}
`;

const StyledInvalidText = styled.Text`
  margin: 0px 2px;
  font-size: 12px;
  font-weight: 500;
  margin-top: 5px;
  color: #ff8989;
`;

interface Props {
  testID?: string;
  errorTestID?: string;
  type?: string;
  isRow?: boolean;
  style?: ViewStyle;
  label?: string;
  labelTextStyle?: TextStyle;
  labelWidth?: number;
  value?: TextInputProps['value'];
  contentStyle?: ViewStyle;
  inputContainerType?: string;
  inputContainerRadius?: number;
  borderStyle?: ViewStyle;
  borderWidth?: number;
  borderColor?: string;
  textStyle?: TextStyle;
  placeholder?: TextInputProps['placeholder'];
  placeholderTextColor?: TextInputProps['placeholderTextColor'];
  secureTextEntry?: TextInputProps['secureTextEntry'];
  onChangeText?: TextInputProps['onChangeText'];
  onSubmitEditing?: (
    e: NativeSyntheticEvent<TextInputSubmitEditingEventData>,
  ) => void;
  leftElement?: ReactElement;
  leftElementStyle?: ViewStyle;
  rightElement?: ReactElement;
  rightElementStyle?: ViewStyle;
  focusedLabelStyle?: TextStyle;
  focusedBorderWidth?: number;
  focusColor?: string;
  errorColor?: string;
  autoCapitalize?: TextInputProps['autoCapitalize'];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  textInputProps?: TextInputProps | any;
  onFocus?: () => void;
  onBlur?: () => void;
  errorText?: string;
  errorTextStyle?: TextStyle;
  keyboardType?: KeyboardTypeOptions;
  numberOfLines?: number;
  multiline?: boolean;
}

export enum EditTextInputType {
  DEFAULT = 'default',
  ROW = 'row',
  BOX = 'box',
  ROW_BOX = 'rowBox',
}

const EditText: FC<Props> = (props) => {
  const [focused, setFocus] = useState(false);

  const {
    testID,
    errorTestID,
    type = EditTextInputType.DEFAULT,
    style,
    label,
    labelTextStyle,
    labelWidth = 110,
    value,
    contentStyle,
    borderStyle,
    borderWidth = 0.6,
    borderColor = '#eaeaf9',
    textStyle = Platform.select({
      ios: {
        paddingVertical: 16,
      },
      android: {
        paddingVertical: 10,
      },
    }),
    placeholder,
    placeholderTextColor,
    secureTextEntry,
    onChangeText,
    onSubmitEditing,
    rightElement,
    rightElementStyle,
    focusedLabelStyle = { fontWeight: 'bold' },
    focusedBorderWidth = 1,
    focusColor = '#79B3F5',
    errorColor = '#FF8989',
    autoCapitalize = 'none',
    textInputProps,
    onFocus,
    onBlur,
    errorText,
    errorTextStyle,
    keyboardType,
    numberOfLines,
    multiline = false,
  } = props;

  switch (type) {
    case EditTextInputType.DEFAULT:
    default:
      return (
        <Container style={style}>
          <StyledLabel
            style={[
              labelTextStyle,
              errorText
                ? { color: errorColor }
                : focused && [
                  { color: focusColor },
                  focusedLabelStyle,
                ],
            ]}
          >
            {label}
          </StyledLabel>
          <StyledTextInput
            {...textInputProps}
            testID={testID}
            autoCapitalize={autoCapitalize}
            onFocus={(): void => {
              setFocus(true);
              onFocus && onFocus();
            }}
            onBlur={(): void => {
              setFocus(false);
              onBlur && onBlur();
            }}
            placeholder={placeholder}
            placeholderTextColor={placeholderTextColor}
            value={value}
            style={textStyle}
            numberOfLines={numberOfLines}
            onChangeText={onChangeText}
            secureTextEntry={secureTextEntry}
            onSubmitEditing={onSubmitEditing}
            multiline={multiline}
          />
          {
            borderWidth && <StyledLine
              style={[
                borderStyle,
                { borderBottomWidth: borderWidth, borderColor: borderColor },
                errorText
                  ? { borderColor: errorColor }
                  : focused && [
                    { borderColor: focusColor },
                    { borderBottomWidth: focusedBorderWidth },
                  ],
              ]}
            />
          }
          {errorText ? (
            <StyledInvalidText
              testID={errorTestID}
              style={[{ color: errorColor }, errorTextStyle]}>
              {`${errorText}`}
            </StyledInvalidText>
          ) : null}
        </Container>
      );
    case EditTextInputType.ROW:
      return (
        <Container style={style}>
          <StyledRowContent
            style={[
              { borderColor: borderColor, borderBottomWidth: borderWidth },
              borderStyle,
              errorText
                ? {
                  borderColor: errorColor,
                  borderBottomWidth: focusedBorderWidth,
                }
                : focused && {
                  borderColor: focusColor,
                  borderBottomWidth: focusedBorderWidth,
                },
              contentStyle,
            ]}
          >
            {label ? (
              <StyledRowLabel
                style={[
                  labelTextStyle,
                  errorText
                    ? [{ color: errorColor }, focusedLabelStyle]
                    : focused && [
                      { color: focusColor },
                      focusedLabelStyle,
                    ],
                  { width: labelWidth },
                ]}
              >
                {label}
              </StyledRowLabel>
            ) : null}
            <StyledRowInput
              {...textInputProps}
              testID={testID}
              style={[textStyle, { textAlign: 'left' }]}
              autoCapitalize={autoCapitalize}
              onFocus={(): void => {
                setFocus(true);
                if (onFocus) {
                  onFocus();
                }
              }}
              onBlur={(): void => {
                setFocus(false);
                if (onBlur) {
                  onBlur();
                }
              }}
              onSubmitEditing={onSubmitEditing}
              placeholder={placeholder}
              placeholderTextColor={placeholderTextColor}
              value={value}
              numberOfLines={numberOfLines}
              onChangeText={onChangeText}
              secureTextEntry={secureTextEntry}
              keyboardType={keyboardType}
              multiline={multiline}
            />
          </StyledRowContent>
          {errorText ? (
            <StyledInvalidText testID={errorTestID} style={errorTextStyle}>
              {errorText}
            </StyledInvalidText>
          ) : null}
        </Container>
      );

    case EditTextInputType.BOX:
      return (
        <Container style={style}>
          <StyledLabel
            style={[
              labelTextStyle,
              errorText
                ? { color: errorColor }
                : focused && [
                  { color: focusColor },
                  focusedLabelStyle,
                ],
            ]}
          >
            {label}
          </StyledLabel>
          <StyledRowContent
            style={[
              {
                borderWidth: borderWidth,
                borderColor: borderColor,
              },
              borderStyle,
              errorText
                ? { borderColor: errorColor, borderWidth: focusedBorderWidth }
                : focused && {
                  borderColor: focusColor,
                  borderWidth: focusedBorderWidth,
                },
              contentStyle,
            ]}
          >
            <StyledTextInput
              {...textInputProps}
              testID={testID}
              autoCapitalize={autoCapitalize}
              onFocus={(): void => {
                setFocus(true);
                if (onFocus) {
                  onFocus();
                }
              }}
              onBlur={(): void => {
                setFocus(false);
                if (onBlur) {
                  onBlur();
                }
              }}
              placeholder={placeholder}
              placeholderTextColor={placeholderTextColor}
              value={value}
              style={textStyle}
              onChangeText={onChangeText}
              secureTextEntry={secureTextEntry}
              onSubmitEditing={onSubmitEditing}
              multiline={multiline}
            />
            {rightElement && (
              <StyledIcon style={[{ width: 40 }, rightElementStyle]}>
                {rightElement}
              </StyledIcon>
            )}

          </StyledRowContent>
          {errorText ? (
            <StyledInvalidText
              testID={errorTestID}
              style={[
                {
                  color: errorColor,
                },
                errorTextStyle,
              ]}
            >
              {`${errorText}`}
            </StyledInvalidText>
          ) : null}
        </Container>
      );

    case EditTextInputType.ROW_BOX:
      return (
        <Container style={style}>
          <StyledRowContent
            style={[
              {
                borderWidth: borderWidth,
                borderColor: borderColor,
              },
              borderStyle,
              errorText
                ? { borderColor: errorColor, borderWidth: focusedBorderWidth }
                : focused && {
                  borderColor: focusColor,
                  borderWidth: focusedBorderWidth,
                },
            ]}
          >
            {label ? (
              <StyledRowLabel
                style={[
                  labelTextStyle,
                  errorText
                    ? [{ color: errorColor }, focusedLabelStyle]
                    : focused
                      ? [{ color: focusColor }, focusedLabelStyle]
                      : null,
                  { marginLeft: 15, width: labelWidth },
                ]}>
                {label}
              </StyledRowLabel>
            ) : null}
            <StyledRowInput
              {...textInputProps}
              testID={testID}
              style={[{ paddingRight: 15 }, textStyle]}
              autoCapitalize={autoCapitalize}
              onFocus={(): void => {
                setFocus(true);
                if (onFocus) {
                  onFocus();
                }
              }}
              onBlur={(): void => {
                setFocus(false);
                if (onBlur) {
                  onBlur();
                }
              }}
              onSubmitEditing={onSubmitEditing}
              placeholder={placeholder}
              placeholderTextColor={placeholderTextColor}
              value={value}
              numberOfLines={numberOfLines}
              onChangeText={onChangeText}
              secureTextEntry={secureTextEntry}
              keyboardType={keyboardType}
              multiline={multiline}
            />
          </StyledRowContent>
          {errorText ? (
            <StyledInvalidText testID={errorTestID} style={errorTextStyle}>
              {errorText}
            </StyledInvalidText>
          ) : null}
        </Container>
      );
  }
};

export { EditText };
