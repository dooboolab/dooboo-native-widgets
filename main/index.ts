export * from './Accordion';
export * from './Button';
export * from './ButtonGroup';
export * from './EditText';
export * from './LoadingIndicator';
export * from './Slider';
export * from './SwitchToggle';
export * from './Select';
