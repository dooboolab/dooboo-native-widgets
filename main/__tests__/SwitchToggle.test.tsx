import React, { ReactElement } from 'react';

import { SwitchToggle } from '../../main';
import { TouchableOpacity } from 'react-native';
import renderer from 'react-test-renderer';

const createTestProps = (
  obj?: Record<string, unknown>,
): Record<string, unknown> => ({
  ...obj,
});

let props;
let component: ReactElement;
// let testingLib: RenderResult;

describe('[SwitchToggle]', (): void => {
  props = createTestProps({
    testID: 'SWITCH_ID',
    onPress: jest.fn(),
  });

  component = <SwitchToggle {...props} />;

  it('should render without crashing', (): void => {
    const rendered = renderer.create(component);

    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('[SwitchToggle] Interaction', (): void => {
    it('should simulate onPress', (): void => {
      const rendered = renderer.create(component);
      const switchToggle = rendered.root.findByType(TouchableOpacity);

      jest.useFakeTimers();

      renderer.act(() => {
        switchToggle.props.onPress();
      });

      jest.runAllTimers();
      expect(props.onPress).toHaveBeenCalled();
    });

    it('should toggle switchOn on press', () => {
      // eslint-disable-next-line @typescript-eslint/no-shadow
      const props = {
        switchOn: false,
        onPress: jest.fn(),
      };

      component = <SwitchToggle {...props} />;

      const rendered = renderer.create(component);
      const switchToggle = rendered.root.findByType(TouchableOpacity);

      jest.useFakeTimers();

      renderer.act(() => {
        switchToggle.props.onPress();
      });

      jest.runAllTimers();
      expect(props.switchOn).toBeFalsy();

      // renderer.act(() => {
      //   switchToggle.props.onPress();
      // });
      // expect(props.switchOn).toBeTruthy();

      renderer.act(() => {
        switchToggle.props.onPress();
      });

      expect(props.switchOn).toBeFalsy();
    });
  });
});
