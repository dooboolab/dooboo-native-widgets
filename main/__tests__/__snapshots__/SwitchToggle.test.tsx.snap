// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SwitchToggle] should render without crashing 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="SWITCH_ID"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": undefined,
        "flexDirection": "row",
      }
    }
  >
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "backgroundColor": undefined,
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <View>
        <Text />
      </View>
    </View>
    <View>
      <Text />
    </View>
    <View
      style={
        Object {
          "position": "absolute",
          "right": 5,
        }
      }
    />
  </View>
</View>
`;
